CONVENTIONS.md

```diff
@@ -0,0 +1,125 @@
+# Style conventions
+
+## Docstrings
+
+Docstrings shall be written in NumpyDoc format.
+
+Docstrings can either be just the summary text, or they must be complete.
+
+## Type annotations
+
+Type annotations will not be applied (for now).
+
+Note: I'm not very familiar with the latest and greatest type annotation
+options for libraries that make heavy use of numpy arrays like this one.
+
+## Linting
+
+Linting shall be performed with flake8, flake8-isort, flake8-black and pep8-naming.
+
+Automated formatting shall be performed with black and isort.
+
+Black is left at defaults, flake8 and isort are configured to adhere to black.
+
+## Testing
+
+All functions need to be covered by unit tests.
+
+
+# Functional API conventions
+
+This API is for internal use and for power-users that want to
+vectorize operations on large sets of things.
+
+Performance is prioritized over extensive input validation.
+
+The source for this API resides in the `pylinalg.func` subpackage and is organized
+by type.
+
+## Function naming
+
+The functional API has rather verbose names, but it makes things
+explicit.
+
+## Function signatures
+
+We strive to align closely with the conventions of [numpy](https://numpy.org/doc/stable/reference/ufuncs.html#ufuncs-kwargs)
+in order to be least-surprising for pylinalg users.
+
+All functions shall adhere as much possible to the following signature:
+
+`func(x1, x2, ..., /, *, out=None, dtype=None, **kwargs)`
+
+* All input arguments are positional-only.
+* There can be as many input arguments as needed for the function signature to make sense.
+* Input arguments are always [`array_like`](https://numpy.org/doc/stable/glossary.html?highlight=array_like#term-array_like):
+    * "Any scalar or sequence that can be interpreted as an ndarray. In
+      addition to ndarrays and scalars this category includes lists
+      (possibly nested and with different element types) and tuples.
+      Any argument accepted by `np.array` is `array_like`."
+    * TBD: require a call to `np.asarray` or `np.asanyarray` on inputs?
```

Comment:
I am against `np.asanyarray`; it allows object arrays and you usually don't want these in a numerics library. `np.asarray` or `np.ascontiguousarray` make sense in different places, but I wouldn't require it on each function.

For future reference, I like this SO answer as an overview: https://stackoverflow.com/a/52103839/6753182

Reply:
* I agree with `np.asarray` (and not `np.asanyarray`).
* Thanks for the SO reference. I know the differences, I'm just trying to reach consensus with you two on a decision. Thanks for the feedback, it helps make the right decision.

"`np.asarray` make sense in different places, but I wouldn't require it on each function." -> this is where the struggle lies; what is a good convention? When does it make sense, and when does it not make sense?

In my experience, it just gets added whenever people have the desire to call a function with a plain list instead of an ndarray, and the implementation just "accidentally" happens to perform an operation on that specific input that does not already implicitly call `np.asarray`.

Reply:
FWIW `np.asarray` does not make a copy if the input already is an array. So it should not matter too much performance-wise.

Reply:
> Thanks for the SO reference. I know the differences, I'm just trying to reach consensus with you two on a decision. Thanks for the feedback, it helps make the right decision.

Sorry, that may have come across differently from what I intended. I was thinking of this reference as a useful tool to discuss when we want to use which variant :)

> This is where the struggle lies; what is a good convention? When does it make sense, and when does it not make sense?

I agree with @almarklein here, that it is probably safe to add in most cases, since it is indeed a no-op if the array already is a base `ndarray` or a view and it is indeed an easy way to support ArrayLike input.

At the same time, there are two scenarios that come to mind where `np.asarray` is counter-productive. First, when it is important to preserve array subtypes, e.g. for sparse arrays, masked arrays, or if you have metadata (like ImageIOs old `image` class). Second, if you have an object whose buffer you could share with the base array, ie., where it is sufficient to create a view into a buffer instead of a dedicated array (I would have to check numpy's [`__array__` magic](https://numpy.org/devdocs/user/basics.dispatch.html) again to read up on if this can be solved elegantly).

Both cases can be handled separately though, so I would still opt for using `np.asarray` by default. That said, I would probably not make this a requirement in the conventions or at least state that there may be exceptions for special circumstances (like the above).